### ISSUE TITLE
Playbook ends "NO MORE HOSTS LEFT" after successful run_once task

### ISSUE THREAD

**Issue by mdedek** [ansible/ansible#83581]
### Summary

Task after "run_once" task is not performed when first host failed somewhere earlier.
Playbook ends with message "NO MORE HOSTS LEFT" which doesn't make sense.

### Issue Type

Bug Report

### Component Name

run_once

### Ansible Version

```console
$ ansible --version
ansible [core 2.16.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/ma069836_01/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.12/site-packages/ansible
  ansible collection location = /usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.12.1 (main, Feb 21 2024, 10:25:11) [GCC 8.5.0 20210514 (Red Hat 8.5.0-21)] (/usr/bin/python3.12)
  jinja version = 3.1.2
  libyaml = True

$ ansible --version
ansible [core 2.15.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/ma069836_01/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.11/site-packages/ansible
  ansible collection location = /usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.11.2 (main, Jun  6 2023, 07:39:01) [GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] (/usr/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = /etc/ansible/ansible.cfg
DEFAULT_EXECUTABLE(/etc/ansible/ansible.cfg) = /bin/bash
DEFAULT_FORCE_HANDLERS(/etc/ansible/ansible.cfg) = True
DEFAULT_HOST_LIST(/etc/ansible/ansible.cfg) = ['/etc/ansible/no_hosts']
DEFAULT_LOG_PATH(/etc/ansible/ansible.cfg) = /var/log/ansible.log
DEFAULT_MANAGED_STR(/etc/ansible/ansible.cfg) = This file is managed by Ansible, all changes will be lost !!!
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False

CONNECTION:
==========

paramiko_ssh:
____________
host_key_checking(/etc/ansible/ansible.cfg) = False

ssh:
___
host_key_checking(/etc/ansible/ansible.cfg) = False
```

### OS / Environment

RHEL 8.8
RHEL 8.10

### Steps to Reproduce

```yaml (paste below)
all:
  hosts:
    host1:
    host2:
    host3:
   
---
- hosts: all
  connection: local
  gather_facts: false

  tasks:
    - name: Fail on "host1"
      fail:
        msg: "hostname matches"
      when: inventory_hostname == "host1"

    - name: Print Message 1 only once
      debug:
        msg: "Message 1 only once"
      run_once: True

    - name: Print Message 2
      debug:
        msg: "Message 2"

```

### Expected Results

Final task "Print Message 2" should be performed for remaining host2 a and host3.

PLAY [all] ***********************************************************************************************************

TASK [Fail on "host1"] ***********************************************************************************************
fatal: [host1]: FAILED! => {"changed": false, "msg": "hostname matches"}
skipping: [host2]
skipping: [host3]

TASK [Print Message 1 only once] *************************************************************************************
ok: [host2] => {
    "msg": "Message 1 only once"
}

TASK [Print Message 2] *************************************************************************************
ok: [host2] => {
    "msg": "Message 2"
}
ok: [host3] => {
    "msg": "Message 2"
}

PLAY RECAP ***********************************************************************************************************
host1                      : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
host2                      : ok=2    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
host3                      : ok=1    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0

### Actual Results

```console
PLAY [all] ***********************************************************************************************************

TASK [Fail on "host1"] ***********************************************************************************************
fatal: [host1]: FAILED! => {"changed": false, "msg": "hostname matches"}
skipping: [host2]
skipping: [host3]

TASK [Print Message 1 only once] *************************************************************************************
ok: [host2] => {
    "msg": "Message 1 only once"
}

NO MORE HOSTS LEFT ***************************************************************************************************

PLAY RECAP ***********************************************************************************************************
host1                      : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
host2                      : ok=1    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
host3                      : ok=0    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/plugins/strategy/__init__.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/__init__.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by mdedek** [comment]
This issue may be related to https://github.com/ansible/ansible/issues/81533 but in my opinion it is not limited to 'block'.

**Comment by mkrizek** [comment]
> DEFAULT_FORCE_HANDLERS(/etc/ansible/ansible.cfg) = True

I'll note that the above is important for reproducing the issue.

> This issue may be related to https://github.com/ansible/ansible/issues/81533 but in my opinion it is not limited to 'block'.

This is indeed a duplicate of said issue, or rather the issue has been fixed by fe94a99aa291d129aa6432e5d50e7117d9c6aae3. The fix is included in ansible-core `2.17`.

We'll potentially look at backporting the fix into ansible-core `2.16`.

**Comment by s-hertel** [comment]
Closing as per above
